### PR TITLE
Add build constraints to allow buildling subctl with only non-deploy …

### PR DESCRIPTION
**Pull request by maayanf24**
…commands

Related to: #https://github.com/submariner-io/enhancements/issues/196

### CODE REVIEW

File: pkg/cloud/aws/aws.go [hunk 1]
```diff
@@ -1,3 +1,5 @@
+//go:build !non_deploy
```
Comment (Jaanki) [inline pkg/cloud/aws/aws.go]: No need to add constraints in this file. Applies to all files under pkg/. Only files under cmd/subctl should have the constraint.

**Review by Jaanki — CHANGES_REQUESTED** [review]
Constraining only the cobra related files will do the trick. i.e add constraints only to files under `cmd/subctl/*` and not `pkg/*`

**Review by skitt — CHANGES_REQUESTED** [review]
Please add a system test: build the non-deploy `subctl`, and use it to run `show`, `diagnose` etc.

Please also shorten the commit message:

> Add build constraints for a non-deploy subctl

**Comment by maayanf24** [comment]
> Please add a system test: build the non-deploy `subctl`, and use it to run `show`, `diagnose` etc.
> 
> Please also shorten the commit message:
> 
> > Add build constraints for a non-deploy subctl

After discussing with @Jaanki I only added test for `non_deploy_subctl show all` to save the time it would take to also run `diagnose`, `gather` and the other commands

**Comment by skitt** [comment]
> After discussing with @Jaanki I only added test for `non_deploy_subctl show all` to save the time it would take to also run `diagnose`, `gather` and the other commands

I see you also added sanity tests to make sure that non-deploy `subctl` really can’t `deploy-broker` and `join`, nice!

**Review by skitt — APPROVED** [review]
I’m looking into the linting error.

File: Makefile [hunk 2]
```diff
@@ -125,6 +125,9 @@ cmd/bin/subctl-%: $(shell find . -name "*.go")
 		-X 'github.com/submariner-io/submariner-operator/api/v1alpha1.DefaultRepo=$(DEFAULT_REPO)'" \
 	$(SCRIPTS_DIR)/compile.sh $@ ./cmd
 
+cmd/bin/non_deploy_subctl:
+	$(GO) build -o cmd/bin/non_deploy_subctl --tags non_deploy ./cmd
```
Comment (Jaanki) [inline Makefile]: Add LDFLAGS like its done at L123 above.

File: scripts/test/system.sh [hunk 3]
```diff
@@ -198,6 +206,13 @@ _subctl benchmark throughput "${KUBECONFIGS_DIR}"/kind-config-cluster1 "${KUBECO
 
 # Test subctl cloud prepare invocations
 
+# test non deploy subctl for deploy commands
+_non_deploy_subctl deploy-broker --kubeconfig "${KUBECONFIGS_DIR}"/kind-config-cluster1 && exit 1
+_non_deploy_subctl join --kubeconfig "${KUBECONFIGS_DIR}"/kind-config-cluster1 broker-info.subm --clusterid cluster1 && exit 1
```
Comment (Jaanki) [inline scripts/test/system.sh:211]: Add similar tests for all ignored commands. That way, we don't miss any. In future, if the build constraint is accidentally removed from some commands, this job should fail.

**Review by Jaanki — CHANGES_REQUESTED** [review]

**Comment by skitt** [comment]
The linting error is [an upstream bug](https://github.com/denis-tingaikin/go-header/issues/18), I think the only thing we can do (short of using another tool to check headers) is to ignore files with a build constraint. At the end of `.golangci.yml`, add an `issues.exclude-rules` rule for the `goheaders` linter on the relevant paths.

File: Makefile [hunk 4]
```diff
@@ -125,6 +125,14 @@ cmd/bin/subctl-%: $(shell find . -name "*.go")
 		-X 'github.com/submariner-io/submariner-operator/api/v1alpha1.DefaultRepo=$(DEFAULT_REPO)'" \
 	$(SCRIPTS_DIR)/compile.sh $@ ./cmd
 
+cmd/bin/non_deploy_subctl:
+	mkdir -p cmd/bin
```
Comment (skitt) [inline Makefile]: Let’s use some Make variables to avoid repeating stuff:

```suggestion
	mkdir -p $(@D)
```

File: Makefile [hunk 5]
```diff
@@ -125,6 +125,14 @@ cmd/bin/subctl-%: $(shell find . -name "*.go")
...
+	$(GO) build -o cmd/bin/non_deploy_subctl --tags non_deploy \
```
Comment (skitt) [inline Makefile]:
```suggestion
	$(GO) build -o $@ --tags non_deploy \
```

File: .golangci.yml [hunk 6]
```diff
@@ -201,3 +201,8 @@ issues:
     - path: internal/(cli|env|log)/.*\.go
       linters:
         - goheader
+
+    # Ignore header linting for build constraint
```
Comment (skitt) [inline .golangci.yml:205]:
```suggestion
    # Ignore header linting for build constraint
    # See https://github.com/denis-tingaikin/go-header/issues/18
```

**Review by skitt — CHANGES_REQUESTED** [review]
Sorry, a few Makefile tweaks I didn’t think of earlier, and a pointer to the goheader bug pushing us to add golangci-lint exclusions.

File: Makefile (hunk 4 above)
Comment (maayanf24) [reply Makefile]: Should I update L116 as well?

Comment (skitt) [reply Makefile]: Ah yes, let’s do that, thanks!

**Review by skitt — APPROVED** [review]

**Comment by skitt** [comment]
@Jaanki can you re-review?

**Review by Jaanki — APPROVED** [review]

**Review by tpantelis — APPROVED** [review]

**Pull request merged**